locking: Add helpers for rwsem and osq locks

imran-kn opened the pull request:
These changes add helpers for rwsem and osq locks.
Also change existing helpers when needed

imran-kn commented:
I am also working on spinlock helpers and fall back mechanism for lock scanners (including the ones that have already been merged)when frame['var'] method fails to locate the lock. But this will take some more time .In the mean time , these helpers which have been ready for some time can be reviewed and merged . Once the spinlock helpers and other enhancements are ready I will create another PR for those

brenns10 reviewed (changes requested):
I only got part of the way through this, I mostly focused on the osq helpers, and will need to read more about their application to the mutex and rwsem code after. My comments here are almost entirely nits or small suggestions, it's looking great so far, thank you!

imran-kn commented on drgn_tools/lock.py:
```diff
@@ -128,6 +216,26 @@ def scan_sem_lock(prog: Program, stack: bool) -> None:
         show_sem_lock(prog, frame_list, seen_sems, stack)
 
 
+def scan_rwsem_lock(prog: Program, stack: bool) -> None:
+    """Scan for read-write(rw) semphores"""
+    seen_rwsems: Set[int] = set()
+    frame_list = bt_has(prog, "__rwsem_down_write_failed_common")
+    if frame_list:
+        show_rwsem_lock(prog, frame_list, seen_rwsems, stack)
+
+    frame_list = bt_has(prog, "__rwsem_down_read_failed_common")
+    if frame_list:
+        show_rwsem_lock(prog, frame_list, seen_rwsems, stack)
+
+    frame_list = bt_has(prog, "rwsem_down_write_slowpath")
+    if frame_list:
+        show_rwsem_lock(prog, frame_list, seen_rwsems, stack)
+
+    frame_list = bt_has(prog, "rwsem_down_read_slowpath")
+    if frame_list:
+        show_rwsem_lock(prog, frame_list, seen_rwsems, stack)
```
True. I also found it unoptimal while testing vmcores locally. bt_has was taking few secs but I went ahead with what was available at that point of time. I had some optimisations in mind, mentioning them here as you are already working on it:
1. For scanning waiters we just need to traverse tasks that are in blocked state, may be possible once my other PR for traversing task in different states is available
2. have a version of bt_has that can be used with a given task_struct. that way we could cache tasks that are waiting on X and check from remaining the ones waiting on Y

brenns10 replied:
The biggest issue is that multiple calls to `bt_has()` will re-run the search. So the biggest performance win would be a `bt_has_any(prog, ["function1", "function2"])` rather than calling `bt_has()` several times. In fact, my current optimization simply caches a mapping of each function name to the tasks which contain it, for all functions that are on any stack. This isn't much more expensive than searching for a single function, and it makes repeated calls cheaper.

From there, your optimizations could be helpful as well. I pushed my suggestions in #72, let me know what you think and I'll continue looking at this.

imran-kn replied:
Sounds good. I will check and update you. thanks for starting on it quickly , will be a big help for sure.

imran-kn commented on drgn_tools/lock.py:
```diff
@@ -25,18 +25,79 @@
+                for cpu in cpu_list:
+                    run_time_ns = get_current_run_time(prog, cpu)
+                    run_time_s = nanosecs_to_secs(run_time_ns)
+                    print(
+                        f"CPU: {cpu} has been spinning for {run_time_s} seconds."
+                    )
```
Actually these are not spinlocks. And even for spinlocks as long as someone is spinning , he is spinning with preemption disabled so he should not move away from the CPU . So if the spinners spins for 100 ms and does that 20 times, the above snippet will show only its spin time since it last arrival on CPU. It will not show cumulative spin time. But I get your point about the message. It has scope of being misconstrued :) .How about "has been spinning for X secs, since it last got on CPU" is more appropriate than "not scheduled" since spinner is already on CPU.
These osq spinners are also spinning on a per-cpu variable , similar logic applies to them as well.

brenns10 replied:
I started this review thinking of OSQ spinners as the spinlocks and wrote this comment with that in mind, but I realized that was wrong and never updated it... sorry!  (I guess it's the qspinlocks whose code looks very similar to the OSQ spinners. 😫 )

In any case I also worded my complaint badly and it was a bit pedantic... your suggestion looks great.

imran-kn replied:
No probs @brenns10, indeed qspinlock code shares similarity and its easy to mistake one for other.I have my spinlock waiters helper based on similar logic, but recently with a UEK-5 bug I found it was not showing up all waiters. Will create a PR for that separately. Thanks for the reviews so far :)

brenns10 commented on drgn_tools/locking.py:
```diff
@@ -85,3 +229,238 @@ def for_each_mutex_waiter(prog: Program, mutex: Object) -> Iterable[Object]:
+    for waiter in for_each_rwsem_waiter_entity(rwsem):
+        waiter_type = get_rwsem_waiter_type(waiter)
+        task = waiter.task
+        print(
+            f"  ({task.type_.type_name()})0x{task.value_():x}  (pid){task.pid.value_():<8}  (waiter_type){waiter_type:<16}  (cpu){task_cpu(task):<4}  (state){task_state_to_char(task):<4} (wait_time){timestamp_str(task_lastrun2now(task))}"
+        )
```
> `({task.type_.type_name()})0x{task.value_():x}`

It looks like you're formatting something like `(struct task_struct *)0xffffdeadbeef`. You can do that a bit easier with:
```
>>> task.format_(dereference=False)
'(struct task_struct *)0xffff897a60da9d00'
```

That being said, that is a very long line of output, and it looks like you're printing a table, just... without the column headers, and you're using the parentheses to mark each field. Maybe this is just a difference of opinion, but I think it's a bit cleaner to just represent things as a table.

```
TASK    PID      TYPE   CPU  ST   WAIT
0xfoo     1 down_read     1   R    20s
```

With the table helpers, we can even handle the case where you want to to a backtrace in between each row, as there are two implementations: `drgn_tools.table.Table`, and `drgn_tools.table.FixedTable`. The latter prints each row as it comes in (with fixed column widths), while the former buffers them all and prints them at the end, when it can determine appropriate column widths for all rows. So you could do something like:

```python
# need FixedTable so each row is printed immediately
tbl = FixedTable(["TASK:>x", "PID:>", "TYPE:16s", "CPU:>", "ST", "WAIT:>"])
for waiter in for_each_rwsem_waiter_entity(rwsem):
    waiter_type = get_rwsem_waiter_type(waiter)
    task = waiter.task
    tbl.row(
        task.value_(),
        task.pid.value_(),
        waiter_type,
        task_cpu(task),
        task_state_to_char(task),
        timestamp_str(task_lastrun2now(task)),
    )
    if callstack:
        bt(task)
tbl.write()  # for compatibility with regular Table API
```

imran-kn replied:
I had totally missed the fact that we have these nice table helpers now in drgn_tools. IIRC these came in when Richard was working on dentry helpers. Using table is definitely much better option here and I will integrate your suggestion.

brenns10 commented on drgn_tools/locking.py:
```diff
@@ -85,3 +229,238 @@ def for_each_mutex_waiter(prog: Program, mutex: Object) -> Iterable[Object]:
+    # commit 64489e78004c ('locking/rwsem: Implement a new locking scheme')
+    # (also since v5.3.1) removed usage of different BIAS(es) and re-defined
+    # usage and interpretation of ->count bits.
+    # So although type change of ->owner and re-definition of ->count bits
+    # happened in 2 different commits, both of these changes are available
+    # since v5.3.1.
+    # So we can use ->owner type to distinguish between new and old usage
+    # of rwsem ->count and ->owner bits.
```
Nice!

brenns10 commented on drgn_tools/locking.py:
```diff
@@ -85,3 +229,238 @@ def for_each_mutex_waiter(prog: Program, mutex: Object) -> Iterable[Object]:
+    """
+
+    prog = rwsem_waiter.prog_
+    if rwsem_waiter.type.value_() == prog["RWSEM_WAITING_FOR_WRITE"].value_():
+        waiter_type = "down_write"
+    elif rwsem_waiter.type.value_() == prog["RWSEM_WAITING_FOR_READ"].value_():
+        waiter_type = "down_read"
```
Tiny nitpick, these can be done with `prog.constant()`, which I slightly prefer now that I've done the CTF / kallsyms implementation, as that allows us to skip the symbol table lookup and go straight into the hash table lookup for CTF enum values.

It's probably a negligible actual runtime impact, of course.

brenns10 commented on drgn_tools/util.py:
```diff
@@ -441,3 +441,22 @@ def timestamp_str(ns: int) -> str:
+    """
+
+    prog = val.prog_
+    for cpu in for_each_possible_cpu(prog):
+        if per_cpu(prog[name], cpu).value_() == val.value_():
```
Maybe we could pull `prog[name]` out of the loop, so we don't repeat that lookup several times?

```
var = per_cpu(prog[name], cpu)
````

imran-kn replied:
Done

brenns10 reviewed (changes requested):
A few more small comments, the helpers are looking good. Nice job detecting the rwsem differences via a commit in the same release, that's a great way to do it!

imran-kn commented on drgn_tools/locking.py:
```diff
@@ -85,3 +229,238 @@ def for_each_mutex_waiter(prog: Program, mutex: Object) -> Iterable[Object]:
+    """
+    prog = rwsem.prog_
+    if not rwsem.count.counter.value_():
+        print("rwsem is free.")
+        return NULL(prog, "struct task_struct *")
+
+    if is_rwsem_writer_owned(rwsem):
+        if (rwsem.owner.type_.type_name() != "atomic_long_t") and (
+            rwsem.owner.value_() & _RWSEM_ANONYMOUSLY_OWNED
+        ):
+            print("rwsem is owned by anonymous writer")
```
The idea behind keeping these prints was that when needed we can check owner of any standalone or embedded rwsem , from drgn shell or cli, using this helper and if the owner can't be obtained because of either of the above reasons, we get to know the reason. I have found this helpful when working side by side with crash and drgn , where sometime I locate a rwsem in crash and want to know who owns it. This is why I intend to keep these. Please let me know your thoughts.

brenns10 replied:
I see. I agree that this makes most sense for doing interactive debugging, if you directly call the function, then getting the message printed on the command line makes sense.

If this is used as part of a more complex helper, then it may be a bit confusing to see a message printed out which is not part of the helper. For example, suppose we had a helper that printed a table of rwsems. Then the printout wouldn't be very helpful, because this line gets printed in between each row of the table.

What if we could make a slightly more complex return type, that expresses everything. For example:

```python
import collections
import enum

class RwsemStateCode(enum.Enum):
    UNLOCKED = "unlocked"
    READER_OWNED = "reader owned"
    WRITER_OWNED = "writer owned"
    WRITER_ANONYMOUSLY_OWNED = "writer anonymously owned"

class RwsemState(collections.namedtuple):
    state: RwsemStateCode
    owner: drgn.Object
    """A ``struct rwsem_waiter *``, non-NULL for WRITER_OWNED"""
```

Then we could have a single helper that can return all the necessary details. With that, you would be able to see on the command line that the rwsem was anonymously owned, but it wouldn't interfere with other helpers using that helper. In fact, it would make that info available to those helpers as well!

This is just an example, but I think something inspired by this might be good. I'm not opposed to having the print statements in drgn-tools, but I'm not confident they'll be acceptable in drgn upstream (I'm starting to look at that PR as well).

imran-kn replied:
I agree @brenns10 and have modified get_rwsem_owner now. Interactive users should use get_rwsem_info but if they are just interested in owner info they can still use get_rwsem_owner as per its description. Let me know how it looks now. And thanks for looking into the upstream PR as well :)

brenns10 replied:
I checked https://github.com/oracle-samples/drgn-tools/pull/69/commits/cd9c0e2dfc1f7d6b6b80e832dc06528e3cd4092e and that's great! Just what I was thinking and I think it makes the code a bit clearer.

Just double checking, is this ready for review and merge? I will take a deep look if so.

imran-kn replied:
@brenns10 yes its ready for review and merge now

imran-kn commented on drgn_tools/lock.py:
```diff
@@ -25,18 +25,79 @@
+from drgn_tools.task import nanosecs_to_secs
+
+
+def scan_osq_node(prog: Program, verbosity=0) -> None:
```
Done

imran-kn commented on drgn_tools/lock.py:
```diff
@@ -25,18 +25,79 @@
+                         till the point of spin
+    """
+
+    osq_spinners: defaultdict = defaultdict(list)
```
As it turned out mypy was bothered about annotation. Done now.

imran-kn commented on drgn_tools/lock.py:
```diff
@@ -25,18 +25,79 @@
+                prog,
+                "struct optimistic_spin_node",
+                address=osq_node.next.value_(),
+            )
```
Thanks for pointing this out. I have gone ahead with osq_node.next[0]

imran-kn commented on drgn_tools/lock.py:
```diff
@@ -116,6 +177,33 @@ def show_sem_lock(prog: Program, frame_list, seen_sems, stack: bool) -> None:
 
 
+def show_rwsem_lock(
+    prog: Program, frame_list, seen_rwsems, stack: bool
```
Done

imran-kn commented on drgn_tools/locking.py:
```diff
@@ -6,26 +6,170 @@
+    while osq_node.prev and osq_node.prev.next == osq_node.address_of_():
+        osq_node = Object(
+            prog, "struct optimistic_spin_node", address=osq_node.prev.value_()
+        )
```
Done

imran-kn commented:
> A few more small comments, the helpers are looking good. Nice job detecting the rwsem differences via a commit in the same release, that's a great way to do it!

Thank @brenns10  ! I have addressed review comments received so far. I have  left one comment about prints unresolved. Could  you please have a look and let me know if the reasoning sounds good to you.

brenns10 commented:
Imran, thanks! I've given a suggestion for the print statements. Let me know how you feel and I'll do a final round of review.

brenns10 reviewed (approved):
This looks really excellent, thank you for these!

brenns10 commented:
Oh, unfortunately there were some changes to the `drgn_tools/lock.py` file which need some attention. There was a `pid` and `time` option added to each of the scanning functions to filter the output. I think you'll need to merge or rebase on the latest `main` branch. Sorry for that.

imran-kn commented:
> Oh, unfortunately there were some changes to the `drgn_tools/lock.py` file which need some attention. There was a `pid` and `time` option added to each of the scanning functions to filter the output. I think you'll need to merge or rebase on the latest `main` branch. Sorry for that.

No probs @brenns10 . I have rebased wrt to main branch and no conflicts now.

brenns10 reviewed (dismissed):
I want to merge this as-is. However, there's a follow-up item we'll need to deal with. The merge conflicts may be resolved, but the corelens `lock` command now has `--pid` and `--time` filters, which this code doesn't implement.

The problem is that the rwsem doesn't have clear owners -- at least not in the anonymous/reader owned case. We could implement some filtering by `pid`, but it would miss reader owners. The same would go for the `time` filter as a result.

I think the minimum cleanup we'd want to do is something like:

```python
if pid or time:
    print("warning: skipping rwsem, since reliable owner info cannot be determined")
else:
    scan_rwsem_lock(...)
```

imran-kn commented:
> I want to merge this as-is. However, there's a follow-up item we'll need to deal with. The merge conflicts may be resolved, but the corelens `lock` command now has `--pid` and `--time` filters, which this code doesn't implement.
> 
> The problem is that the rwsem doesn't have clear owners -- at least not in the anonymous/reader owned case. We could implement some filtering by `pid`, but it would miss reader owners. The same would go for the `time` filter as a result.
> 
> I think the minimum cleanup we'd want to do is something like:
> 
> ```python
> if pid or time:
>     print("warning: skipping rwsem, since reliable owner info cannot be determined")
> else:
>     scan_rwsem_lock(...)
> ```

Thanks @brenns10 and yes I looked at pid and time based filtering but my idea was that since this PR was created before the introduction of pid and time based filtering and these filters are optional, I will get this PR merged and can add these filters in a separate PR. So that I don't introduce new work in an already reviewed code.
Let me know if this sounds good to you.

Now coming to pid (and time) based filtering, since pid based filtering is being done for waiters we should be fine for reader owned or anonymously owned rwsems  also.

Having said that (and admittedly I should have raised this earlier), I am not sure if in current scanners , pid based filtering is happening properly. As per my understanding the pid passed to scan mutex lock should be a known mutex/sem waiter but if I already know that this pid is a mutex/sem waiter, should I invoke scan_xyz_lock for this pid. On the other hand if say pid: X is not a mutex/sem waiter but someone passed it to scan_mutex/sem_lock then else condition of show_mutex/sem_lock would still show X as one of the waiters.

This doubt was also another reason for me currently staying away from pid based filtering.

My idea of pid based filtering is that someone passes pids to scan_xyz_lock (without knowing before hand that these pids are lock waiter because for known waiters scanning will not help much) and then if scan_xyz_lock  or any of the subroutines find that one or more of these pids are indeed waiters then they dump the pids and locks (on which these pids are blocked and different pids can be blocked on different locks) info.

This is how I am doing it for rwsem waiters. Please let me know if it makes sense or if I have misundersttos some part of current pid based filtering.

Also since all lock waiters will be in the respective lists in the same order in which they tried to grab the lock, they will be sorted by wait time by default. So if we print the waiter list from tail we will be able to see waiters beyond threshold wait time.

imran-kn commented:
> > I want to merge this as-is. However, there's a follow-up item we'll need to deal with. The merge conflicts may be resolved, but the corelens `lock` command now has `--pid` and `--time` filters, which this code doesn't implement.
> > The problem is that the rwsem doesn't have clear owners -- at least not in the anonymous/reader owned case. We could implement some filtering by `pid`, but it would miss reader owners. The same would go for the `time` filter as a result.
> > I think the minimum cleanup we'd want to do is something like:
> > ```python
> > if pid or time:
> >     print("warning: skipping rwsem, since reliable owner info cannot be determined")
> > else:
> >     scan_rwsem_lock(...)
> > ```
> 
> Thanks @brenns10 and yes I looked at pid and time based filtering but my idea was that since this PR was created before the introduction of pid and time based filtering and these filters are optional, I will get this PR merged and can add these filters in a separate PR. So that I don't introduce new work in an already reviewed code. Let me know if this sounds good to you.
> 
> Since pid based filtering is happening for waiters we are fine for anonymously owned or reader owned rwsems too. If its okay I can add the change in this PR or create a separate one as mentioned earlier

@brenns10 anyways I went ahead and added pid and time based filtering as well. Now that  box is ticked. So if you could just review [923ea5ddf8cc](https://github.com/oracle-samples/drgn-tools/pull/69/commits/923ea5ddf8cc9cc34ed6a1f2457bf7b157295a4f) , we should be ready to merge this
